signac/contrib/_index.py

```diff
@@ -0,0 +1,312 @@
+
+        # Extract all logical-operator expressions for now.
+        or_expressions = expr.pop("$or", None)
+        and_expressions = expr.pop("$and", None)
+        not_expression = expr.pop("$not", None)
+
+        # Reduce the result based on the remaining non-logical expression:
+        for key, value in _nested_dicts_to_dotted_keys(expr):
+            reduce_results(self._find_expression(key, value))
+            if not result_ids:
+                # No matches, so exit early.
+                return set()
+
+        # Reduce the result based on the logical-operator expressions:
+        if not_expression is not None:
+            not_match = self._find_result(not_expression)
+            reduce_results(set(self).difference(not_match))
+            if not result_ids:
+                # No matches, so exit early.
+                return set()
+
+        if and_expressions is not None:
+            _check_logical_operator_argument("$and", and_expressions)
+            for expr_ in and_expressions:
+                reduce_results(self._find_result(expr_))
+                if not result_ids:
+                    # No matches, so exit early.
+                    return set()
+
+        if or_expressions is not None:
+            _check_logical_operator_argument("$or", or_expressions)
+            or_results = set()
+            for expr_ in or_expressions:
+                or_results.update(self._find_result(expr_))
+            reduce_results(or_results)
+
+        return result_ids
+
+    def find(self, filter_=None):
+        """Return a result vector of ids for the given filter and limit.
+
+        This function normalizes the filter argument and then attempts to
+        build a result vector for the given key-value queries.
+        For each key that is queried, an internal index is built and then
+        searched.
+
+        The results are a set of ids, where each id is the value of the
+        primary key of a document that matches the given filter.
+
+        The find() method uses the following optimizations:
+
+            1. If the filter is None, the result is directly returned, since
+               all documents will match an empty filter.
+            2. The filter is processed key by key, once the result vector is
+               empty it is immediately returned.
+
+        Parameters
+        ----------
+        filter_ : dict
+            The filter argument that all documents must match (Default value = None).
+
+        Returns
+        -------
+        set
+            A set of ids of documents that match the given filter.
+
+        Raises
+        ------
+        ValueError
+            When the filter argument is invalid.
+
+        """
+        if not filter_:
+            return set(self)
+
+        filter_ = json.loads(json.dumps(filter_))  # Normalize
+        if type(filter_) is not dict:
+            raise ValueError(filter_)
+        return set(self._find_result(filter_))
```

Comment:
I think it's already always a `set`, right? In case there's a code path that I'm missing, can we change that?
```suggestion
        return self._find_result(filter_)
```